kuksa_databroker/databroker/src/grpc/kuksa_val_v1/val.rs

```diff
@@ -54,55 +55,141 @@ impl proto::val_server::Val for broker::DataBroker {
-                                reason: "not_found".to_owned(),
-                                message: "Path not found".to_owned(),
-                            }),
-                        });
-                    }
-                    Err(ReadError::PermissionExpired) => {
-                        errors.push(proto::DataEntryError {
-                            path: request.path,
-                            error: Some(proto::Error {
-                                code: 401,
-                                reason: "unauthorized".to_owned(),
-                                message: "Authorization expired".to_owned(),
-                            }),
-                        });
+                match glob::matches_path_pattern(&request.path) {
+                    true => {
+                        match broker.get_entry_by_path(&request.path).await {
+                            Ok(entry) => {
+                                let view =
+                                    proto::View::from_i32(request.view).ok_or_else(|| {
+                                        tonic::Status::invalid_argument(format!(
+                                            "Invalid View (id: {}",
+                                            request.view
+                                        ))
+                                    })?;
+                                let fields = HashSet::<proto::Field>::from_iter(
+                                    request.fields.iter().filter_map(
+                                        |id| proto::Field::from_i32(*id), // Ignore unknown fields for now
+                                    ),
+                                );
+                                let fields = combine_view_and_fields(view, fields);
+                                debug!("Getting fields: {:?}", fields);
+                                let proto_entry = proto_entry_from_entry_and_fields(entry, fields);
+                                debug!("Getting datapoint: {:?}", proto_entry);
+                                entries.push(proto_entry);
+                            }
+                            Err(ReadError::NotFound) => {
+                                let sub_path_new = request.path.clone();
+                                let regex = glob::to_regex_new(sub_path_new.as_ref());
+                                match broker
+                                    .get_entries_by_wildcards_with_regex(&regex.unwrap())
+                                    .await
+                                {
+                                    Ok(entries_result) => {
+                                        for data_entry in entries_result {
+                                            if request.view == 2
+                                                && data_entry.metadata.entry_type
+                                                    == broker::EntryType::Actuator
+                                            {
```

Comment:
This is really weird?

```rust
if request.view == 2
    && data_entry.metadata.entry_type
        == broker::EntryType::Actuator
{
    // ... do stuff
} else if request.view != 2 {
    // ... do stuff
}
```

First off, `... do stuff` is _exactly_ the same in both branches (at least I think so).

Secondly, I don't know what `request.view == 2` means (by looking at it, I will now look it up the the proto/generated file). Instead, use the generated enum `proto::View::*` for these comparisons, which you get access to on the very next line:
```rust
let view = proto::View::from_i32(request.view)
```

Thirdly, by doing these conditional things based on `request.view` will probably lead to weird/erroneous behaviour. Requesting View::VIEW_TARGET_VALUE should be the equivalent of requesting Field::FIELD_ACTUATOR_TARGET (most probably).

Reply:
That was initially implemented with the intention to cover the 'getTargetValue' which provides the view with FIELD_ACTUATOR_TARGET value to just get actuators.
And yes, it should be refined, there is a repeated code.